### PR TITLE
add tutorial of support ast restore to sql text 

### PR DESCRIPTION
**Pull request by leoppro**
See: pingcap/tidb#8532 , pingcap/tidb#8533

### CODE REVIEW

File: support-ast-restore-to-sql-text.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    	// Name is the referenced column name.
+    	Name *ColumnName
+    
+    	// Refer is the result field the column name refers to.
+    	// The value of Refer.Expr is used as the value of the expression.
+    	Refer *ResultField
+    }
+    ```
+    
+    我们发现我们需要先实现 ColumnName 的 Restore 函数：
```
[the remark this answers is not in the document]
Comment (leoppro) [reply support-ast-restore-to-sql-text.md]: 看`parser.y`文件3376行，parse过程并没有对Refer赋值

[the remark this answers is not in the document]
Comment (leoppro) [reply support-ast-restore-to-sql-text.md]: 这地方我加个说明

**Comment by leoppro** [comment]
@CaitinChen @XuHuaiyu PTAL

**Comment by leoppro** [comment]
@shenli  PTAL

File: support-ast-restore-to-sql-text.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+SQL 语句发送到 TiDB 后首先会经过 parser，从文本 parse 成为 AST（抽象语法树），AST 节点与 SQL 文本结构是一一对应的，我们通过遍历整个 AST 树就可以拼接出一个与 AST 语义相同的 SQL 文本。
+
+对 parser 不熟悉的小伙伴们可以看 [TiDB 源码阅读系列文章（五）TiDB SQL Parser 的实现](https://www.pingcap.com/blog-cn/tidb-source-code-reading-5/)。
+
+我们当前 AST 结构定义如下（以 `ast.CreateUserStmt` 为例）：
+
+![create-user-stmt](media/create-user-stmt.png)
+
+我们在 `ast.Node` 接口中添加了一个 `Restore(sb *strings.Builder) error` 函数，这个函数将当前节点对应的 SQL 文本追加至参数 `sb` 中，如果节点结构无效则返回 `error`。
```
Comment (zz-jason) [inline support-ast-restore-to-sql-text.md]:
```suggestion
我们在 `ast.Node` 接口中添加了一个 `Restore(sb *strings.Builder) error` 函数，这个函数将当前节点对应的 SQL 文本追加至参数 `sb` 中，如果节点无效则返回 `error`。
```

**Comment by leoppro** [comment]
@XuHuaiyu @zz-jason PTAL again

**Comment by shenli** [comment]
LGTM

File: support-ast-restore-to-sql-text.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
+---
+title: 十分钟成为 Contributor 系列 | 支持 AST 还原为 SQL
+author: ['赵一霖']
+date: 2018-12-05
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]: Please remember to update the date.

File: support-ast-restore-to-sql-text.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    ...
+}
+```
+
+以 SQL 语句 `SELECT column0 FROM table0 UNION SELECT column1 FROM table1 WHERE a = 1` 为例，如下图所示，我们通过遍历整个 AST 树，递归调用每个节点的 `Restore()` 方法，即可拼接成一个完整的 SQL 文本。
+
+![ast-tree](media/ast-tree.png)
+
+值得注意的是，SQL 文本与 AST 是一个多对一的关系，我们不可能从 AST 结构中还原出与原 SQL 完全一致的文本，
+因此我们只要保证还原出的 SQL 文本与原 SQL **语义相同** 即可。所谓语义相同，指的是由 AST 还原出的 SQL 文本再被解析为 AST 后，两颗 AST 是相等的。
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
因此我们只要保证还原出的 SQL 文本与原 SQL **语义相同** 即可。所谓语义相同，指的是由 AST 还原出的 SQL 文本再被解析为 AST 后，两个 AST 是相等的。
```

Comment (CaitinChen) [reply support-ast-restore-to-sql-text.md]: 用“棵”也好像太生动形象了 哈哈

File: support-ast-restore-to-sql-text.md [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+我们已经完成了接口设计和测试框架，具体的`Restore()` 函数留空。因此**只需要选择一个留空的 `Restore()` 函数实现，并添加相应的测试数据，就可以提交一个 PR 了！**
+
+## **实现 `Restore()` 函数的整体流程**
+
+0. 请先阅读 [Proposal](https://github.com/pingcap/tidb/tree/master/docs/design/2018-11-29-ast-to-sql-text.md)、[Issue](https://github.com/pingcap/tidb/issues/8532)
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
1. 请先阅读 [Proposal](https://github.com/pingcap/tidb/tree/master/docs/design/2018-11-29-ast-to-sql-text.md)、[Issue](https://github.com/pingcap/tidb/issues/8532)
```

File: support-ast-restore-to-sql-text.md [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+1. 在 [Issue](https://github.com/pingcap/tidb/issues/8532) 中找到未实现的函数
+
+    在 [Issue-pingcap/tidb#8532](https://github.com/pingcap/tidb/issues/8532) 中找到一个没有被其他贡献者认领的任务，例如 `ast/expressions.go: BetweenExpr`。
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    1. 在 [Issue-pingcap/tidb#8532](https://github.com/pingcap/tidb/issues/8532) 中找到一个没有被其他贡献者认领的任务，例如 `ast/expressions.go: BetweenExpr`。
```

File: support-ast-restore-to-sql-text.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+    
+    在 [pingcap/parser](https://github.com/pingcap/parser) 中找到任务对应文件 `ast/expressions.go`。
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    2. 在 [pingcap/parser](https://github.com/pingcap/parser) 中找到任务对应文件 `ast/expressions.go`。
```

File: support-ast-restore-to-sql-text.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    
+    在文件中找到 `BetweenExpr` 结构的 `Restore` 函数：
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    3. 在文件中找到 `BetweenExpr` 结构的 `Restore` 函数：
```

File: support-ast-restore-to-sql-text.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+    ```
+    // Restore implements Node interface.
+    func (n *BetweenExpr) Restore(ctx *RestoreCtx) error {
+        return errors.New("Not implemented")
+    }
+    ```
+
+2. 实现 `Restore()` 函数
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
3. 实现 `Restore()` 函数
```

File: support-ast-restore-to-sql-text.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+
+    根据 Node 节点结构和 SQL 语法实现函数功能。
+    
+     > 参考 [MySQL 5.7 SQL Statement Syntax](https://dev.mysql.com/doc/refman/5.7/en/sql-syntax.html)
+
+3. 写单元测试
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
4. 写单元测试
```

File: support-ast-restore-to-sql-text.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+
+    参考示例在相关文件下添加单元测试。
+
+4. 运行 `make test`，确保所有的 test case 都能跑过。
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
5. 运行 `make test`，确保所有的 test case 都能跑过。
```

File: support-ast-restore-to-sql-text.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+
+5. 提交 PR
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
6. 提交 PR
```

File: support-ast-restore-to-sql-text.md [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+     PR 标题统一为：`parser: implement Restore for XXX`  
+     请在 PR 中关联 Issue: `pingcap/tidb#8532`
+
+## **示例**
+
+这里以[实现 BetweenExpr 的 Restore 函数 PR](https://github.com/pingcap/parser/pull/71/files) 为例，进行详细说明：
+
+1. 首先看 `ast/expressions.go`：
+
+    我们要实现一个 `ast.Node` 结构的 `Restore` 函数，首先清楚该结构代表什么短语，例如 `BetweenExpr` 代表 `expr [NOT] BETWEEN expr AND expr` (参见：[MySQL 语法 - 比较函数和运算符](https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_between))。
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    1. 我们要实现一个 `ast.Node` 结构的 `Restore` 函数，首先清楚该结构代表什么短语，例如 `BetweenExpr` 代表 `expr [NOT] BETWEEN expr AND expr` (参见：[MySQL 语法 - 比较函数和运算符](https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_between))。
```

File: support-ast-restore-to-sql-text.md [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+    
+    观察 `BetweenExpr` 结构：
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    2. 观察 `BetweenExpr` 结构：
```

File: support-ast-restore-to-sql-text.md [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        // AND 左侧的表达式
+        Left ExprNode
+        // AND 右侧的表达式
+        Right ExprNode
+        // 是否有 NOT 关键字
+        Not bool
+    }
+    ```
+    
+    实现 `BetweenExpr` 的 `Restore` 函数：
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    3. 实现 `BetweenExpr` 的 `Restore` 函数：
```

File: support-ast-restore-to-sql-text.md [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+        }
+        // Run Test
+        RunNodeRestoreTest(c, testCases, "select %s", extractNodeFunc)
+    }
+    ```
+    
+    **至此 `BetweenExpr` 的 `Restore` 函数实现完成，可以提交 PR 了。为了更好的理解测试逻辑，下面我们看 `RunNodeRestoreTest`：**
+    
+    ```
+    // 下面是测试逻辑，已经实现好了，不需要贡献者实现
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    // 下面是测试逻辑，已经实现好了，不需要 contributor 实现
```

Comment (CaitinChen) [reply support-ast-restore-to-sql-text.md]: “contributor” is also used in the Chinese blogs.

File: support-ast-restore-to-sql-text.md [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+        // 是否能被正确 parse
+        ok      bool
+        // 预期的 restore SQL
+        restore string
+    }
+    ```
+    
+    测试代码会判断原 SQL parse 出 AST 后再还原的 SQL 是否与预期的 restore SQL 相等，具体的测试逻辑在 `parser_test.go` 中 `RunTest()`、`RunRestoreTest()` 函数，逻辑与前例类似，此处不再赘述。
+
+编辑按：添加 TiDB Robot 微信，加入 TiDB Contributor Club，无门槛参与开源项目，改变世界从这里开始吧（萌萌哒）。
```
Comment (CaitinChen) [inline support-ast-restore-to-sql-text.md]:
```suggestion
编者按：添加 TiDB Robot 微信，加入 TiDB Contributor Club，无门槛参与开源项目，改变世界从这里开始吧（萌萌哒）。
```

File: support-ast-restore-to-sql-text.md [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+
+// RestoreCtx is `Restore` context to hold flags and writer.
+type RestoreCtx struct {
+	Flags RestoreFlags
+	In    io.Writer
+}
+
+// WriteKeyWord 用于向 `ctx` 中写入关键字（例如：SELECT）。
+// 它的大小写受 `RestoreKeyWordUppercase`，`RestoreKeyWordLowercase` 控制
+func (ctx *RestoreCtx) WriteKeyWord(keyWord string) {
```
Comment (XuHuaiyu) [inline support-ast-restore-to-sql-text.md]: Please add the close brace

File: support-ast-restore-to-sql-text.md [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+    }
+    ```
+    
+    测试代码会判断原 SQL parse 出 AST 后再还原的 SQL 是否与预期的 restore SQL 相等，具体的测试逻辑在 `parser_test.go` 中 `RunTest()`、`RunRestoreTest()` 函数，逻辑与前例类似，此处不再赘述。
+
+---
+
+添加 TiDB Robot 微信，加入 TiDB Contributor Club，无门槛参与开源项目，改变世界从这里开始吧（萌萌哒）。
+
+![](media/tidb-robot.jpg "tidb_rpbot")
```
Comment (husiyu) [inline support-ast-restore-to-sql-text.md]: This picture is better 👉 https://upload-images.jianshu.io/upload_images/542677-6952db2eff44ffe8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240

File: support-ast-restore-to-sql-text.md [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+        // 预期的 restore SQL
+        restore string
+    }
+    ```
+    
+    测试代码会判断原 SQL parse 出 AST 后再还原的 SQL 是否与预期的 restore SQL 相等，具体的测试逻辑在 `parser_test.go` 中 `RunTest()`、`RunRestoreTest()` 函数，逻辑与前例类似，此处不再赘述。
+
+---
+
+添加 TiDB Robot 微信，加入 TiDB Contributor Club，无门槛参与开源项目，改变世界从这里开始吧（萌萌哒）。
```
Comment (husiyu) [inline support-ast-restore-to-sql-text.md]:
```suggestion
加入 TiDB Contributor Club，无门槛参与开源项目，改变世界从这里开始吧（萌萌哒）。
```

File: support-ast-restore-to-sql-text.md [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+     PR 标题统一为：`parser: implement Restore for XXX`  
+     请在 PR 中关联 Issue: `pingcap/tidb#8532`
+
+## **示例**
+
+这里以[实现 BetweenExpr 的 Restore 函数 PR](https://github.com/pingcap/parser/pull/71/files) 为例，进行详细说明：
+
+1. 首先看 `ast/expressions.go`：
+
+    1. 我们要实现一个 `ast.Node` 结构的 `Restore` 函数，首先清楚该结构代表什么短语，例如 `BetweenExpr` 代表 `expr [NOT] BETWEEN expr AND expr` (参见：[MySQL 语法 - 比较函数和运算符](https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_between))。
```
Comment (husiyu) [inline support-ast-restore-to-sql-text.md]:
```suggestion
    1. 我们要实现一个 `ast.Node` 结构的 `Restore` 函数，首先清楚该结构代表什么短语，例如 `BetweenExpr` 代表 `expr [NOT] BETWEEN expr AND expr` （参见：[MySQL 语法 - 比较函数和运算符](https://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#operator_between)）。
```

**Comment by husiyu** [comment]
LGTM

**Pull request merged**